finality-aleph/src/justification/requester.rs

```diff
@@ -202,6 +213,67 @@ where
     }
 
     pub fn finalized_number(&self) -> NumberFor<B> {
-        self.client.info().finalized_number
+        self.backend.blockchain().info().finalized_number
+    }
+
+    fn request(&mut self, header: <B as BlockT>::Header) {
+        let number = *header.number();
+        let hash = header.hash();
+        debug!(target: "aleph-justification",
+               "We have block {:?} with hash {:?}. Requesting justification.", number, hash);
+        self.justification_request_scheduler.on_request_sent();
+        self.block_requester.request_justification(&hash, number);
+    }
+
+    // We request justifications for all the children of last finalized block and a justification
+    // for a block of number num on longest branch.
+    // Assuming that we request at the same pace that finalization is progressing, the former ensures
+    // that we are up to date with finalization. On the other hand, the latter enables fast catch up
+    // if we are behind.
+    // We also request the child that it's on the same branch as top_wanted since a fork may happen
+    // somewhere in between them.
+    fn request_targets(&mut self, mut top_wanted: NumberFor<B>) -> Vec<<B as BlockT>::Header> {
+        let blockchain_backend = self.backend.blockchain();
+        let blockchain_info = blockchain_backend.info();
+        let finalized_hash = blockchain_info.finalized_hash;
+
+        let mut targets = blockchain_backend
+            .children(finalized_hash)
+            .unwrap_or_default()
+            .into_iter()
+            .filter_map(|hash| {
+                if let Ok(Some(header)) = blockchain_backend.header(BlockId::Hash(hash)) {
+                    Some(header)
+                } else {
+                    warn!(target: "aleph-justification",
+                      "Cancelling request for child {:?} of {:?}: no block.", hash, finalized_hash);
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        self.request_status
+            .save_children(finalized_hash, targets.len());
+        let best_number = blockchain_info.best_number;
+        if best_number <= top_wanted {
+            // most probably block best_number is not yet finalized
+            top_wanted = best_number - NumberFor::<B>::one();
+        }
+        match blockchain_backend.header(BlockId::Number(top_wanted)) {
+            Ok(Some(header)) => {
+                if !targets.contains(&header) {
```

Comment:
I think we don't need this `contains`. We can just push and `save_block` anyway.

Reply:
We can but there is an easy `O(1)` check so I used it.